Source/Scene/ModelExperimental/InstancingPipelineStage.js

```diff
@@ -0,0 +1,272 @@
+import Cartesian3 from "../../Core/Cartesian3.js";
+import ComponentDatatype from "../../Core/ComponentDatatype.js";
+import defined from "../../Core/defined.js";
+import Matrix4 from "../../Core/Matrix4.js";
+import AttributeType from "../AttributeType.js";
+import Quaternion from "../../Core/Quaternion.js";
+import Buffer from "../../Renderer/Buffer.js";
+import BufferUsage from "../../Renderer/BufferUsage.js";
+import InstanceAttributeSemantic from "../InstanceAttributeSemantic.js";
+import ModelExperimentalUtility from "./ModelExperimentalUtility.js";
+import InstancingStageVS from "../../Shaders/ModelExperimental/InstancingStageVS.js";
+
+/**
+ * The instancing pipeline stage is responsible for handling GPU mesh instancing at the node
+ * level.
+ *
+ * @namespace
+ * @private
+ */
+var InstancingPipelineStage = {};
+
+/**
+ * Process a node. This modifies the following parts of the render resources:
+ * <ul>
+ *  <li>adds attribute declarations for the instancing vertex attributes in the vertex shader</li>
+ *  <li>adds an instancing translation min and max to compute an accurate bounding volume</li>
+ * </ul>
+ * @param {RenderResources.NodeRenderResources} renderResources The render resources for this node.
+ * @param {ModelComponents.Node} node The node.
+ * @param {FrameState} frameState The frame state.
+ */
+InstancingPipelineStage.process = function (renderResources, node, frameState) {
+  var instances = node.instances;
+  var instancingVertexAttributes = [];
+
+  var shaderBuilder = renderResources.shaderBuilder;
+  shaderBuilder.addDefine("HAS_INSTANCING");
+  shaderBuilder.addVertexLines([InstancingStageVS]);
+
+  var translationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.TRANSLATION
+  );
+
+  var translationMax;
+  var translationMin;
+  if (defined(translationAttribute)) {
+    translationMax = translationAttribute.max;
+    translationMin = translationAttribute.min;
+  }
+
+  var rotationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.ROTATION
+  );
+  if (
+    defined(rotationAttribute) ||
+    (!defined(translationMax) && !defined(translationMin))
+  ) {
+    var transformsTypedArray = getInstanceTransformsTypedArray(
+      instances,
+      renderResources
+    );
+    var transformsVertexBuffer = Buffer.createVertexBuffer({
+      context: frameState.context,
+      typedArray: transformsTypedArray.buffer,
+      usage: BufferUsage.STATIC_DRAW,
+    });
+
+    var vertexSizeInFloats = 12;
+    var componentByteSize = ComponentDatatype.getSizeInBytes(
+      ComponentDatatype.FLOAT
+    );
+
+    instancingVertexAttributes = [
+      {
+        index: ++renderResources.attributeIndex,
```

Comment:
Oh one subtlety, this has to be a `++` because POSITION must be index 0 and so everything else must start with index 1

Maybe it would be good to rethink how we do indices for clarity? Like this:

* start the overall count at 1, not 0
* every time you add an attribute, push and then increment
* POSITION is always hardcoded to 0

Reply:
I went back and forth on starting at `1` vs `0`, but your explanation makes sense and I implemented it.